File: programmers/2022_KAKAO/양과 늑대/one.js
function solution(info, edges) {
  var answer = 1;
  const dfs = (
    start = 0,
    state = { sheepes: 1, wolves: 0 },
    cadidates = []
  ) => {
    cadidates = [
      ...cadidates,
      ...edges.filter((e) => e[0] === start).map((e) => e[1]),
    ];

    for (let nextidx in cadidates) {
      let next = cadidates[nextidx];
      let newCandidates = [...cadidates];
      newCandidates.splice(nextidx, 1);

      if (info[next] === 0) {
        answer = Math.max(answer, state.sheepes + 1);
        dfs(next, { ...state, sheepes: state.sheepes + 1 }, newCandidates);
      } else if (state.sheepes > state.wolves + 1) {
        dfs(next, { ...state, wolves: state.wolves + 1 }, newCandidates);
      }
    }
  };

  dfs();
  return answer;
}

const info = [0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1];
const edges = [
  [0, 1],
  [1, 2],
  [1, 4],
  [0, 8],
  [8, 7],
  [9, 10],
  [9, 11],
  [4, 3],
  [6, 5],
  [4, 6],
  [8, 9],
];

console.log(solution(info, edges));

File: programmers/2022_KAKAO/양과 늑대/one.py
answer=0
def solution(info, edges):
    
    graph = [[ ] for i in range(len(info))]
    for p,ch in edges:
        graph[p].append(ch)
        
    def dfs(start = 0, candi=[], sheep=0, wolf=0):
        global answer
        sheep += 1 if info[start]==0 else 0
        wolf += info[start]
        answer = max(answer, sheep)
        
        if wolf >= sheep:
            return
        candi.extend(graph[start])
        for node in candi:
            newCandi = candi[:]
            newCandi.remove(node)
            dfs(node, newCandi,sheep, wolf)
    dfs()
    return answer